packages/mui-material-next/src/ButtonBase/ButtonBase.tsx

```diff
@@ -108,9 +108,16 @@ const ButtonBase = React.forwardRef(function ButtonBase<
     TouchRippleProps,
     touchRippleRef,
     type,
+    hostElementName: hostElementNameProp,
     ...other
   } = props;
 
+  let hostElementName = hostElementNameProp;
+
+  if (other.href || other.to) {
+    hostElementName = 'a';
+  }
+
```

Comment:
I would move [the `ComponentProp` logic](https://github.com/mui/material-ui/pull/40171/files#diff-be027e60d418dd4a195bfd57a4536a8689ae44094da36f4c99d1118e38103a53R133-R136) up here so both are set in the same `if`. That way, we avoid changing one without the other in the future.
```suggestion
  let ComponentProp = component;
  let hostElementName = hostElementNameProp;

  if (ComponentProp === 'button' && (other.href || other.to)) {
    ComponentProp = LinkComponent;
    hostElementName = 'a';
  }

```

Reply:
This makes sense 👍 

I've changed it to:
```diff
- hostElementName = 'a';
+ hostElementName = hostElementNameProp ?? 'a';
```
so that the user-provided `hostElementName` prop will have priority, and will trigger the warning if the `hostElementName` prop doesn't match what is actually rendered, for example:

```jsx
<Button href="https://mui.com/" hostElementName="span">
  Here
</Button>
```
This should render an `a` because of the `href` prop, so `hostElementName="span"` doesn't match and the warning will trigger